tx-indexer/src/storage.rs

```diff
@@ -135,24 +133,66 @@ impl CacheStorage {
             self.push_receipt_to_watching_list(receipt_id.clone(), transaction_key.clone())
                 .await?;
             for outcome in self.storage.get_tx_outcomes(transaction_key).await? {
-                let indexed_outcome = near_indexer_primitives::IndexerExecutionOutcomeWithReceipt {
-                    execution_outcome: outcome.execution_outcome.clone(),
-                    receipt: outcome.receipt.clone(),
+                // Skip the outcome that is already in the transaction
+                if outcome.execution_outcome.id == tx_details.transaction_outcome.id {
+                    continue;
                 };
-                for receipt_id in indexed_outcome.execution_outcome.outcome.receipt_ids.iter() {
+                for receipt_id in outcome.execution_outcome.outcome.receipt_ids.iter() {
                     self.push_receipt_to_watching_list(
                         receipt_id.to_string(),
                         transaction_key.clone(),
                     )
                     .await?;
                 }
-                self.push_outcome_and_receipt(transaction_key, indexed_outcome, false)
+                self.push_outcome_and_receipt_to_cache(transaction_key, outcome)
                     .await?;
             }
         }
         Ok(())
     }
 
+    async fn push_outcome_and_receipt_to_storage(
```

Comment:
Let's add tracing instrumentation here as we have it in other places (docstring explaining the purpose for future reference would be nice, too)